pypiserver/config.py

```diff
@@ -0,0 +1,599 @@
+    """A config for the Run command."""
+
+    def __init__(self, namespace: argparse.Namespace) -> None:
+        """Construct a RuntimeConfig."""
+        super().__init__(namespace)
+        self.port: int = namespace.port
+        self.interface: str = namespace.interface
+        self.authenticate: t.List[str] = namespace.authenticate
+        self.password_file: t.Optional[str] = namespace.passwords
+        self.disable_fallback: bool = namespace.disable_fallback
+        self.fallback_url: str = namespace.fallback_url
+        self.server_method: str = namespace.server
+        self.overwrite: bool = namespace.overwrite
+        self.hash_algo: t.Callable = namespace.hash_algo
+        self.welcome_msg: str = namespace.welcome
+        self.cache_control: t.Optional[int] = namespace.cache_control
+        self.log_req_frmt: str = namespace.log_req_frmt
+        self.log_res_frmt: str = namespace.log_res_frmt
+        self.log_err_frmt: str = namespace.log_err_frmt
+
+
+class UpdateConfig(_ConfigCommon):
+    """A config for the Update command."""
+
+    def __init__(self, namespace: argparse.Namespace) -> None:
+        """Construct an UpdateConfig."""
+        super().__init__(namespace)
+        self.execute: bool = namespace.execute
+        self.download_directory: t.Optional[str] = namespace.download_directory
+        self.allow_unstable: bool = namespace.allow_unstable
+        self.ignorelist: t.List[str] = namespace.ignorelist_file
+
+
+class Config:
+    """Config constructor for building a config from args."""
+
+    @classmethod
+    def from_args(
+        cls, args: t.Sequence[str] = None
+    ) -> t.Union[RunConfig, UpdateConfig]:
+        """Construct a Config from the passed args or sys.argv."""
+        args = args if args is not None else sys.argv[1:]
+        parser = get_parser()
+        try:
+            # The parser prints regardless of whether we're catching the
+            # error or not, so we redirect sdout for this first run.
+            # However, we'll _keep_ the output and print it if the second
+            # run fails, since the first run will give the user feedback
+            # related to what they actually put in.
+            with capture_stderr() as cap:
+                try:
+                    parsed = parser.parse_args(args)
+                finally:
+                    cap.seek(0)
+                    orig_err = cap.read()
+            if parsed.cmd is None:
+                # No command was found, which happens if no arguments were
+                # provided. We will raise a SystemExit so we get the same
+                # deprecated argument handling functionality as we do for
+                # a failed parse.
+                orig_err = None
+                raise SystemExit
+        except SystemExit:
+            # argparse could not parse the arguments. Perhaps they are
+            # old-style arguments, so we'll try adjusting and re-parsing.
+            err_msg = (
```

Comment:
Yeah, normally I'm a stickler for explicit null checks, to avoid unexpected bugs with falsy values, but I think we'd want to use the usage message if this is an empty string anyway, so the falsy check will be good